duckduckgo/Android
# Android: Site-level "fireproof" feature to preserve logins

Task/Issue URL: https://app.asana.com/0/72649045549333/1175609299816961/f
Tech Design URL: 
CC: 

**Description**:
This is a feature branch that will merge into develop following previous PRs:

https://github.com/duckduckgo/Android/pull/796

- fireproof website option menu + confirmation snackbar
- fireproof websites settings option
- fireproof websites screen (list of fireproof websites)
- Room database to store fireproof websites
- improvements on bookmark screen (to keep consitency with fireproof websites)

https://github.com/duckduckgo/Android/pull/808

- Implement logic to remove cookies preserving the ones related to fireproof website.
- Try to directly remove cookies from WebView database, preserving the ones with hosts related to a fireproof website
- If process fails, fallback to remove all the cookies to avoid any leak
- Send pixels in the following scenarios:
- database path not found
- database can't be opened
- delete query fails

https://github.com/duckduckgo/Android/pull/810

- Implement empty state for fireproof websites screen
- Implement feature pixels
- User clicks on "fireproof a website"
- User undo "fireproof website" action (confirmation snackbar after fireproffing a website)
- User removed a website from "fireproof websites"


**Steps to test this PR**:
Listing the most important tests to validate the feature works. For more concrete testing, please go to each PR description.

### Test 1
1. Fresh install
1. Visit any website (e.g: twitter.com)
1. Perform an action that will create cookies (e.g: login)
1. Fireproof website
1. Ensure snackbar appears with expected domain
1. Ensure Fireproof website in overflow menu is now disabled
1. Press fire button
1. Visit fireproof website
1. Ensure state has been preserved (e.g: you are logged in)
1. Open Settings
1. Open Fireproof websites
1. Ensure fireproof website is listed and feature description appears at the top

### Test 2
1. Visit duckduckgo website (do not fireproof DDG site)
1. Set a theme (it will create a cookie)
1. Press fire button
1. Visit duckduckgo again
1. Ensure theme has been preserved

### Test 3
1. Visit any non-fireproof website
1. Open options menu and click on Fireproof website
1. Ensure snackbar appears with expected domain
1. Click on "undo"
1. Ensure Fireproof website in overflow menu is still disabled

### Test 4
1. Without any fireproof website
1. Access settings -> fireproof websites
1. Ensure list is empty and "No websites fireproofed yet" and feature description appears at the top
1. Go back to browser
1. Fireproof any website
1. Navigate to settings -> fireproof websites again
1. Ensure list is not empty and feature description appears at the top
1. Remove all websites from the list
1. Ensure  "No websites fireproofed yet" appeared again

### Test 5
1. checkout develop
1. Fresh install
1. Change to this branch
1. Install the app
1. Ensure database migration correctly executed
1. Perform any of the previous test to check functionality works

Additional test:
- Change between light/dark theme

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)
